cpp/community_detection_module/algorithm/label_propagation.cpp

```diff
@@ -0,0 +1,291 @@
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
+    node_label_Ps[label] = inflated_node_label_Ps;
+    sum_Ps += inflated_node_label_Ps;
+  }
+
+  for (const auto [label, _] : node_label_Ps) node_label_Ps[label] /= sum_Ps;
+
+  return;
+}
+
+void LabelRankT::cutoff(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    double min_value) {
+  std::vector<std::uint64_t> to_be_removed;
+
+  for (const auto [label, P] : node_label_Ps) {
+    if (P < min_value) to_be_removed.push_back(label);
+  }
+
+  for (const auto label : to_be_removed) node_label_Ps.erase(label);
+
+  return;
+}
+
+LabelRankT::LabelRankT(std::unique_ptr<mg_graph::Graph<>>& graph,
+                       double w_selfloop, double similarity_threshold,
+                       double exponent, double min_value)
+    : graph(graph) {
+  this->w_selfloop = w_selfloop;
+  this->similarity_threshold = similarity_threshold;
+  this->exponent = exponent;
+  this->min_value = min_value;
+}
+
+std::pair<bool, std::uint64_t> LabelRankT::iteration(
+    bool incremental, std::unordered_set<std::uint64_t> changed_nodes) {
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      updated_label_Ps;
+
+  bool none_updated = true;
+  std::uint64_t most_updates = 0;
+
+  for (const auto node : graph->Nodes()) {
+    if (incremental) {
+      bool was_updated = changed_nodes.count(node.id) > 0 ? true : false;
+      if (not was_updated) continue;
+    }
+
+    // node selection (a.k.a. conditional update)
+    if (not distinct_enough(node, similarity_threshold)) continue;
+    none_updated = false;
+
+    // label propagation
+    std::unordered_map<std::uint64_t, double> updated_node_label_Ps =
+        propagate(node);
+
+    // inflation
+    inflate(updated_node_label_Ps, exponent);
+
+    // cutoff
+    cutoff(updated_node_label_Ps, min_value);
+
+    updated_label_Ps.insert({node.id, updated_node_label_Ps});
+  }
+
+  for (const auto [node, updated_node_label_Ps] : updated_label_Ps) {
+    label_Ps[node] = updated_node_label_Ps;
+    times_updated[node]++;
+
+    if (times_updated[node] > most_updates) most_updates = times_updated[node];
+  }
+
+  return {none_updated, most_updates};
+}
+
+std::unordered_map<std::uint64_t, std::uint64_t> LabelRankT::get_labels() {
+  std::unordered_map<std::uint64_t, std::uint64_t> labels;
+
+  if (calculated) {
+    for (const auto [node, _] : label_Ps)
+      labels.insert({node, get_label(node)});
+
+    return labels;
+  }
+
+  return calculate_labels();
+}
+
+std::unordered_map<std::uint64_t, std::uint64_t> LabelRankT::calculate_labels(
+    std::uint64_t max_iterations, std::uint64_t max_updates,
+    std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete) {
+  bool incremental = changed_nodes.size() > 1 ? true : false;
+
+  if (incremental) {
```

Comment:
I need explanation for this if-statement

Reply:
If the algorithm is run normally, its data structures are initialized for the entire graph, hence `graph->Nodes()` in the iterator.
With incremental operation, most of the graph is unchanged so the data structure entries are only updated for the changed nodes (and deleted node entries are erased).